pre-commit

```diff
@@ -73,7 +73,7 @@ check() {
         res=$?
         echo "-----------------------"
         echo "$msg: FAILED ($res)"
-        echo "  ${@@Q}"
```

Comment:
I am sorry to be pedantic on a one line PR, but I'm a little confused. The shebang line calls for bash, so why is the the script running in zsh? I also can't reproduce the exit code of zero -- I get an exit code of 1. (I also got a glob failure on `.git/MERGE_*` until I made a file to match it, so that might be worth fixing as well if we really want to run this under bash.

All that aside, it might better to do `" ""$@"`, as I think that's closer to `" ${@@Q}"` than `" ${*}"`.

Reply:
Maybe this issue isn't bash vs zsh... I don't get much other than the "Bad Substitution" error, and a little bit of Googling suggested that `{@@Q}` may not be available in zsh (and my terminal runs zsh, but you raise a good point about the shebang.)

Let me give your suggestion a try.

Reply:
The point behind using `${@@Q}` was to ensure that any quoting needed to have the command-line work was included.  That way, you could have a command with spaces, like `cargo test -- 'something with spaces'` print in a way that could be copied and pasted back.  `"$@"` or `"$*"` don't achieve that.

Of course, if this is failing for someone, that's probably because Apple are shipping an ancient version of bash that predates the move to GPLv3.  That is why new macs tend to use zsh: that old bash is now quite bad.

Reply:
that's probably what's going on. suggestion on the right way to fix it?

Reply:
https://formulae.brew.sh/formula/bash#default perhaps.

Reply:
sure, but it would be nice to have this work out of the box on macOS, no? the fact that it fails but the commit still goes forward seems like a failure case you'd want to prevent as universally as possible.

Reply:
Hmm, it shouldn't.  We have `set -e` on.  That should cause the error to trigger an `exit 1`.

I can't test this easily.  Maybe you can work out how to test for something that isn't broken so that you don't get this failing open like that.

Reply:
ok let me look into it further. it's certainly the worst failure mode and it's important to at least fail correctly

Reply:
this is what I get with `${@@Q}` 😦 
<img width="551" alt="image" src="https://github.com/user-attachments/assets/12d7a06b-6e9a-4d84-9662-3e7c40447088">

but, as you note, with `${*}` or with `" ""$@"`, it drops the quotes, e.g. the command I have in there is `python fail_fast.py --foo "bar"` and it drops those quotes:
<img width="497" alt="image" src="https://github.com/user-attachments/assets/c60183db-4419-4fa1-bdf5-64cfcbb5d112">

Reply:
Sounds like a bug in that version of bash.

If I add a bad substitution, I get:

$ ./pre-commit
-----------------------
fail: FAILED (1)
  'false' '1 2 3'
./pre-commit: line 77:   ${@(}: bad substitution
$ echo $?
1

Reply:
that makes more sense. just pushed a check for the version, and it correctly exits with the stock version of bash.

however, I also updated bash, and what I get out is 
```
$ ./pre-commit
-----------------------
Fail fast: FAILED (1)
 'python' 'fail_fast.py' '--foo' 'bar'
```
where I'd expect
```
python fail_fast.py --foo "bar"
```

Reply:
The quoting is not preserved, but re-added.  That means that you get more quotes than is necessary.  But at least `"foo bar"` doesn't turn into two arguments (`foo bar`) but `foo bar`.